src/core/dyad_core.c

```diff
@@ -0,0 +1,622 @@
+            // TODO Indicate already initialized
+            DPRINTF ((*ctx), "DYAD context already initialized\n");
+        }
+        else
+        {
+            **ctx = dyad_ctx_default;
+        }
+        return DYAD_OK;
+    }
+    printf("Allocating ctx\n");
+    *ctx = (dyad_ctx_t*) malloc(sizeof(struct dyad_ctx));
+    if (*ctx == NULL)
+    {
+        fprintf (stderr, "Could not allocate DYAD context!\n");
+        return DYAD_NOCTX;
+    }
+    printf("Setting ctx values that don't need allocation\n");
+    **ctx = dyad_ctx_default;
+    (*ctx)->debug = debug;
+    (*ctx)->check = check;
+    (*ctx)->shared_storage = shared_storage;
+    (*ctx)->key_depth = key_depth;
+    (*ctx)->key_bins = key_bins;
+    (*ctx)->intercept = intercept;
+    printf("Setting KVS namespace\n");
+    if (kvs_namespace == NULL)
+    {
+        fprintf (stderr, "No KVS namespace provided!\n");
+        // TODO see if we want a different return val
+        return DYAD_NOCTX;
+    }
+    (*ctx)->kvs_namespace = (char*) malloc(strlen(kvs_namespace)+1);
+    if ((*ctx)->kvs_namespace == NULL)
+    {
+        fprintf (stderr, "Could not allocate buffer for KVS namespace!\n");
+        free(*ctx);
+        *ctx = NULL;
+        return DYAD_NOCTX;
+    }
+    strncpy(
+            (*ctx)->kvs_namespace,
+            kvs_namespace,
+            strlen(kvs_namespace)+1
+           );
+    printf("Setting prod_managed_path\n");
+    if (prod_managed_path == NULL)
+    {
+        (*ctx)->prod_managed_path = NULL;
+    }
+    else
+    {
+        (*ctx)->prod_managed_path = (char*) malloc(strlen(prod_managed_path)+1);
+        if ((*ctx)->prod_managed_path == NULL)
+        {
+            fprintf (stderr, "Could not allocate buffer for Producer managed path!\n");
+            free((*ctx)->kvs_namespace);
+            free(*ctx);
+            *ctx = NULL;
+            return DYAD_NOCTX;
+        }
+        strncpy(
+                (*ctx)->prod_managed_path,
+                prod_managed_path,
+                strlen(prod_managed_path)+1
+            );
+    }
+    printf("Setting cons_managed_path\n");
+    if (cons_managed_path == NULL)
+    {
+        (*ctx)->cons_managed_path = NULL;
+    }
+    else
+    {
+        (*ctx)->cons_managed_path = (char*) malloc(strlen(cons_managed_path)+1);
+        if ((*ctx)->cons_managed_path == NULL)
+        {
+            fprintf (stderr, "Could not allocate buffer for Consumer managed path!\n");
+            free((*ctx)->kvs_namespace);
+            free((*ctx)->prod_managed_path);
+            free(*ctx);
+            *ctx = NULL;
+            return DYAD_NOCTX;
+        }
+        strncpy(
+                (*ctx)->cons_managed_path,
+                cons_managed_path,
+                strlen(cons_managed_path)+1
+            );
+    }
+    if ((*ctx)->prod_managed_path == NULL && (*ctx)->cons_managed_path == NULL)
+    {
+        fprintf (stderr, "Warning: DYAD instance is neither producer nor consumer!\n");
+    }
```

Comment:
This check could have happened at the beginning, and skip the initialization other than setting `(*ctx)->intercept = false`